wgengine/netstack/netstack.go

```diff
@@ -290,6 +297,41 @@ func Create(logf logger.Logf, tundev *tstun.Wrapper, e wgengine.Engine, mc *magi
 	} else {
 		linkEP = newLinkEndpoint(512, uint32(tstun.DefaultTUNMTU()), "", disableGRO)
 	}
+	// tcpip.TCP{Receive,Send}BufferSizeRangeOption is gVisor's version of
+	// Linux's tcp_{r,w}mem. Application within gVisor differs as some Linux
+	// features of are not (yet) implemented, and socket buffer memory is not
+	// completely controlled within gVisor, e.g. we allocate
+	// *stack.PacketBuffer's for the write path within Tailscale. Therefore, we
+	// loosen our understanding of the relationship between these Linux and
+	// gVisor tunables, and set a singular value across min, default, and max.
+	// The chosen value is biased towards higher throughput on high
+	// bandwidth-delay product paths.
```

Comment:
Would it be better to go with something lower for devices that are on the memory constrained side, e.g. Android and iOS?

Reply:
Here are some sysctls from a Pixel 6:

```
oriole:/ # uname -a
Linux localhost 5.10.189-android13-4-00012-g1217bb583cc5-ab11174560 #1 SMP PREEMPT Mon Dec 4 18:59:42 UTC 2023 aarch64 Toybox
oriole:/ # sysctl net.ipv4.tcp_rmem
net.ipv4.tcp_rmem = 524288	2097152	8388608
oriole:/ # sysctl net.ipv4.tcp_wmem
net.ipv4.tcp_wmem = 262144	524288	4194304
```

rmem max on this Pixel is 2MiB higher than what we are setting here, wmem max is the same.

Not sure how to check on iOS. iOS (tvOS) would probably benefit a lot here since it's commonly used as an exit node.

It's a loose interpretation in terms of memory usage since there's no concept of min taking effect when under memory pressure, and we allocate stack.PacketBuffer's outside gVisor anyway. Generally speaking max will translate to an increase in memory usage where TCP streams can take advantage. I'm on the fence to go lower for lower mem devices.

Reply:
After further discussion I've excluded iOS from these changes pending further testing, and I've gotten a little more aggressive with all other platforms where there is less of a concern. Some reshuffling of things and force pushed.